src/Microsoft.ML.Transforms/TensorFlow/Tensorflow.cs

```diff
@@ -0,0 +1,1854 @@
+			if (tags == null)
+				throw new ArgumentNullException (nameof (tags));
+			if (exportDir == null)
+				throw new ArgumentNullException (nameof (exportDir));
+			if (metaGraphDef == null)
+				throw new ArgumentNullException (nameof (metaGraphDef));
+			var cstatus = TFStatus.Setup (status);
+			unsafe
+			{
+				var h = TF_LoadSessionFromSavedModel (sessionOptions.handle, runOptions == null ? null : runOptions.LLBuffer, exportDir, tags, tags.Length, graph.handle, metaGraphDef == null ? null : metaGraphDef.LLBuffer, cstatus.handle);
+
+				if (cstatus.CheckMaybeRaise (status)) {
+					return new TFSession (h, graph);
+				}
+			}
+			return null;
+		}
+
+		// extern void TF_CloseSession (TF_Session *, TF_Status *status);
+		[DllImport (NativeBinding.TensorFlowLibrary)]
+		private static extern unsafe void TF_CloseSession (TF_Session session, TF_Status status);
+
+		/// <summary>
+		/// Closes the session.  Contacts any other processes associated with the session, if applicable.
+		/// </summary>
+		/// <param name="status">Status buffer, if specified a status code will be left here, if not specified, a <see cref="T:TensorFlow.TFException"/> exception is raised if there is an error.</param>
+		/// <remarks>
+		/// Can not be called after calling DeleteSession.
+		/// </remarks>
+		public void CloseSession (TFStatus status = null)
+		{
+			if (handle == IntPtr.Zero)
+				ObjectDisposedException ();
+			var cstatus = TFStatus.Setup (status);
+			TF_CloseSession (handle, cstatus.handle);
+			cstatus.CheckMaybeRaise (status);
+		}
+
+		// extern void TF_DeleteSession (TF_Session *, TF_Status *status);
+		[DllImport (NativeBinding.TensorFlowLibrary)]
+		private static extern unsafe void TF_DeleteSession (TF_Session session, TF_Status status);
+
+		/// <summary>
+		/// Deletes the session.
+		/// </summary>
+		/// <param name="status">Status.</param>
+		public void DeleteSession (TFStatus status = null)
+		{
+			if (handle == IntPtr.Zero)
+				ObjectDisposedException ();
+			var cstatus = TFStatus.Setup (status);
+			TF_DeleteSession (handle, cstatus.handle);
+			cstatus.CheckMaybeRaise (status);
+		}
+
+		internal override void NativeDispose (IntPtr handle)
+		{
+			using (var s = new TFStatus ()) {
+				TF_DeleteSession (handle, s.handle);
+			}
+		}
+
+		// extern void TF_SessionRun (TF_Session *session, const TF_Buffer *run_options, const TF_Output *inputs, TF_Tensor *const *input_values, int ninputs, const TF_Output *outputs, TF_Tensor **output_values, int noutputs, const TF_Operation *const *target_opers, int ntargets, TF_Buffer *run_metadata, TF_Status *);
+		[DllImport (NativeBinding.TensorFlowLibrary)]
+		private static extern unsafe void TF_SessionRun (TF_Session session, LLBuffer* run_options, TFOutput [] inputs, TF_Tensor [] input_values, int ninputs, TFOutput [] outputs, TF_Tensor [] output_values, int noutputs, TF_Operation [] target_opers, int ntargets, LLBuffer* run_metadata, TF_Status status);
```

Comment:
Right.  We have to wrap the TFTensor in an ML.NET type and avoid the copy.  It won't be exposed as a managed array, but as some other type that can be accessed like an array.  That wrapping type would need to be disposable.  Is there a type in ML.NET that's a good fit?  I have an appropriate type in S.N.Tensors but we probably want to use whatever type ML.NET standardizes for this.  Looked at VBuffer and it won't work since its a struct that exposes the array, it doesn't have a "pointer" / "IReadOnlyList<T>" mode.  I don't think IDV is correct either.

I looked into C-API to see if we could get TF to directly allocate the resulting tensor as a managed array and it doesn't look like they provide a hook for the allocator, so we'll need to either wrap their memory (and manage its liftime) or copy it.

Let me know your thoughts about a wrapping type for TFTensor or a native pointer.

---
In reply to: [210361853](https://github.com/zeahmed/machinelearning/pull/4#discussion_r210361853) [](ancestors = 210361853)